pkg/storegateway/series_refs.go

```diff
@@ -444,3 +451,218 @@ func (s *deduplicatingSeriesChunkRefsSetIterator) Next() bool {
+	}
+
+	// We can't compute the series hash yet because we're still missing the series labels.
+	// However, if the hash is already in the cache, then we can remove all postings for series
+	// not belonging to the shard.
+	if shard != nil {
+		var unsafeStats queryStats
+		ps, unsafeStats = filterPostingsByCachedShardHash(ps, shard, seriesHashCache)
+		stats.merge(&unsafeStats)
+	}
+
+	postingsIterator := newPostingsSetsIterator(ps, batchSize)
+	inflatingIterator := newLoadingSeriesChunkRefsSetIterator(
+		ctx,
+		postingsIterator,
+		indexr,
+		stats,
+		blockMeta,
+		shard,
+		cachedSeriesHasher{cache: seriesHashCache},
+		skipChunks,
+		minTime,
+		maxTime,
+	)
+
+	limitingIterator := newLimitingSeriesChunkRefsSetIterator(
+		inflatingIterator,
+		chunksLimiter,
+		seriesLimiter,
+	)
+
+	return limitingIterator, nil
+}
+
+func newLoadingSeriesChunkRefsSetIterator(
+	ctx context.Context,
+	postingsSetIterator *postingsSetsIterator,
+	indexr *bucketIndexReader,
+	stats *safeQueryStats,
+	blockMeta *metadata.Meta,
+	shard *sharding.ShardSelector,
+	seriesHasher seriesHasher,
+	skipChunks bool,
+	minTime int64,
+	maxTime int64,
+) *loadingSeriesChunkRefsSetIterator {
+	if skipChunks {
+		minTime, maxTime = blockMeta.MinTime, blockMeta.MaxTime
+	}
+
+	return &loadingSeriesChunkRefsSetIterator{
+		ctx:                 ctx,
+		postingsSetIterator: postingsSetIterator,
+		indexr:              indexr,
+		stats:               stats,
+		blockID:             blockMeta.ULID,
+		shard:               shard,
+		seriesHasher:        seriesHasher,
+		skipChunks:          skipChunks,
+		minTime:             minTime,
+		maxTime:             maxTime,
+	}
+}
+
+func (s *loadingSeriesChunkRefsSetIterator) Next() bool {
+	if s.err != nil {
+		return false
+	}
+	if !s.postingsSetIterator.Next() {
+		return false
+	}
+	nextPostings := s.postingsSetIterator.At()
+	loadedSeries, err := s.indexr.preloadSeries(s.ctx, nextPostings, s.stats)
+	if err != nil {
+		s.err = errors.Wrap(err, "preload series")
+		return false
+	}
+	inflatedSeries := newBucketIndexInflatedSeries(s.blockID, loadedSeries, s.indexr)
+
+	// Track the series loading statistics in a not synchronized data structure to avoid locking for each series
+	// and then merge before returning from the function.
+	loadStats := &queryStats{}
+	defer s.stats.merge(loadStats)
+
+	nextSet := newSeriesChunkRefsSet(len(nextPostings))
+
+	for _, id := range nextPostings {
+		lset, chks, err := inflatedSeries.inflateSeriesForTime(id, s.skipChunks, s.minTime, s.maxTime, loadStats)
+		if err != nil {
+			s.err = errors.Wrap(err, "read series")
+			return false
+		}
+		if lset.Len() == 0 {
+			// No matching chunks for this time duration, skip series
+			continue
+		}
+
+		if !shardOwned(s.shard, s.seriesHasher, id, lset, loadStats) {
+			continue
+		}
+
+		entry := seriesChunkRefs{
+			lset:   lset,
+			chunks: chks,
+		}
+
+		nextSet.series = append(nextSet.series, entry)
```

Comment:
[nit] Tiny simplification

```suggestion
		nextSet.series = append(nextSet.series,  seriesChunkRefs{
			lset:   lset,
			chunks: chks,
		})
```